decompiler/structures/pseudo/complextypes.py

```diff
@@ -0,0 +1,139 @@
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, order=True)
+class ComplexTypeMember(ComplexType):
+    """Class representing a member of a struct type.
+    @param name: name of the struct member
+    @param offset: offset of the member within the struct
+    @param type: datatype of the member
+    @param value: initial value of the member, enums only
+    """
+
+    name: str
+    offset: int
+    type: Type
+    value: Optional[int] = None
+
+    def __str__(self) -> str:
+        return f"{self.name}"
+
+    def declaration(self) -> str:
+        if isinstance(self.type, Union):
+            return self.type.declaration()
+        return f"{self.type.__str__()} {self.name}"
+
+
+@dataclass(frozen=True, order=True)
+class Struct(ComplexType):
+    """Class representing a struct type."""
+
+    members: Dict[int, ComplexTypeMember] = field(compare=False)
+    type_specifier: ComplexTypeSpecifier = ComplexTypeSpecifier.STRUCT
+
+    def add_member(self, member: ComplexTypeMember):
+        self.members[member.offset] = member
+
+    def get_member_by_offset(self, offset: int) -> ComplexTypeMember:
+        return self.members.get(offset)
+
+    def declaration(self) -> str:
+        members = ";\n\t".join(self.members[k].declaration() for k in sorted(self.members.keys())) + ";"
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+
+@dataclass(frozen=True, order=True)
+class Union(ComplexType):
+    members: List[ComplexTypeMember] = field(compare=False)
+    type_specifier = ComplexTypeSpecifier.UNION
+
+    def add_member(self, member: ComplexTypeMember):
+        self.members.append(member)
+
+    def declaration(self) -> str:
+        members = ";\n\t".join(x.declaration() for x in self.members) + ";"
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+    def get_member_by_type(self, _type: Type) -> ComplexTypeMember:
+        """Retrieve member of union by its type."""
+        for member in self.members:
+            if member.type == _type:
+                return member
+
+
+@dataclass(frozen=True, order=True)
+class Enum(ComplexType):
+    members: Dict[int, ComplexTypeMember] = field(compare=False)
+    type_specifier = ComplexTypeSpecifier.ENUM
+
+    def add_member(self, member: ComplexTypeMember):
+        self.members[member.value] = member
+
+    def get_name_by_value(self, value: int) -> str:
+        return self.members.get(value).name
+
+    def declaration(self) -> str:
+        members = ",\n\t".join(f"{x.name} = {x.value}" for x in self.members.values())
+        return f"{self.type_specifier.value} {self.name} {{\n\t{members}\n}}"
+
+
+@dataclass(frozen=True, order=True)
+class ComplexTypeName(Type):
+    """Class that store a name of a complex type. Used to prevent recursions when constructing
+    struct(...) members of the same complex type"""
+
+    name: str
+
+    def __str__(self) -> str:
+        return self.name
```

Comment:
I thought a lot about about that.
Simply having strings there most probably will do the job.
However, I am decided to be paranoid and wrap `str` into a pseudo type so that not after some decompiler stage there are expressions that have `type` `str` instead of subclass of `pseudo.Type`.

Do you have major preference for strings here?